### ISSUE TITLE
Plugin: Strictly type check model state

### ISSUE THREAD

**Issue by ShMcK** [rematch/rematch#221]
It would be great to have a plugin that checks reducer state and validates it against a set of types. This is feature that exists in MobX, but isn't available in Redux (without using TypeScript).

For example:

```js
import { init } from '@rematch/core'
import rematchStateTypingsPlugin from '@rematch/stateTypings'

const count = {
  state: {
    name: '',
    age: null
  },
  typings: {
    name: 'string',
    age: 'number|null'
  },
  reducers: { ... },
}

const store = init({
  models: { count },
  plugins: [rematchStateTypingsPlugin]
})
```

On each state change (in development mode), warnings could be triggered in the console if the next state is not properly validated against the state typings.

This is currently just an idea with a rough outline. I'd love to hear anyone else's thoughts, or whether or not this idea would be useful.

**Comment by Ne0nx3r0** [comment]
I tried to do this using implied typings so you could maintain your actions/reducers and immediately see where things break. 

The issue I kept running into is without either explicitly defining your action types or a fairly unpalatable amount of boilerplate the implied type ends up getting lost and you lose that benefit.

TypeScript has generics and implied types, but I couldn't find a way to define an implied type so you can reference it explicitly without redefining the type.

An alternative for pure Redux is something like this: https://spin.atomicobject.com/2017/07/24/redux-action-pattern-typescript/

But it's quite verbose and falls into that boilerplate problem I mentioned.

**Comment by ShMcK** [comment]
It would definitely be preferable to to validate against a reference to a TypeScript or Flow type.

If there is strong demand for this kind of plugin, I'd consider looking into it.

**Comment by will-stone** [comment]
This reminds me of mobx-state-tree. Typings would certainly add to the power and allure of this package over straight Redux. I’d say go for it if you have the time available to develop it.

**Comment by koss-lebedev** [comment]
Closed in #466
